src/transformers/models/mplugdocowl/modeling_mplugdocowl.py

```diff
@@ -0,0 +1,1994 @@
+        # Transpose to shape: (B, L/X, H)
+
+        sequence_output = torch.cat([sequence_output, self.vit_eos.repeat(B, 1, 1)], dim=1)
+        # Concatenate end-of-sequence token, resulting shape: (B, L/4X + 1, H)
+
+        return sequence_output
+
+
+@add_start_docstrings(
+    """The MPLUGDOCOWL model which consists of a vision backbone and a language model.""",
+    MPLUGDOCOWL_START_DOCSTRING,
+)
+class MPLUGDocOwlForConditionalGeneration(MPLUGDocOwlPreTrainedModel):
+    def __init__(self, config: MPLUGDocOwlConfig):
+        super().__init__(config)
+
+        self.vision_tower = MPLUGDocOwlVisionModel(config.vision_config)
+        self.multi_modal_projector = MPLUGDocOwlHReducer(config)
+        self.vocab_size = config.text_config.vocab_size
+
+        self.language_model = MPLUGDocOwlForCausalLM(config.text_config)
+
+        self.pad_token_id = self.config.pad_token_id if self.config.pad_token_id is not None else -1
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.language_model.get_input_embeddings()
+
+    def set_input_embeddings(self, value):
+        self.language_model.set_input_embeddings(value)
+
+    def get_output_embeddings(self):
+        return self.language_model.get_output_embeddings()
+
+    def set_output_embeddings(self, new_embeddings):
+        self.language_model.set_output_embeddings(new_embeddings)
+
+    def set_decoder(self, decoder):
+        self.language_model.set_decoder(decoder)
+
+    def get_decoder(self):
+        return self.language_model.get_decoder()
+
+    def tie_weights(self):
+        return self.language_model.tie_weights()
+
+    def resize_token_embeddings(self, new_num_tokens: Optional[int] = None, pad_to_multiple_of=None) -> nn.Embedding:
+        model_embeds = self.language_model.resize_token_embeddings(new_num_tokens, pad_to_multiple_of)
+        # update vocab size
+        self.config.text_config.vocab_size = model_embeds.num_embeddings
+        self.vocab_size = model_embeds.num_embeddings
+        return model_embeds
+
+    def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, input_ids, attention_mask, labels):
+        num_images, num_image_patches, embed_dim = image_features.shape
+        batch_size, sequence_length = input_ids.shape
+        left_padding = not torch.sum(input_ids[:, -1] == torch.tensor(self.pad_token_id))
+        # 1. Create a mask to know where special image tokens are
+        special_image_token_mask = input_ids == self.config.image_token_index
+        num_special_image_tokens = torch.sum(special_image_token_mask, dim=-1)
+        # Compute the maximum embed dimension
+        max_embed_dim = (num_special_image_tokens.max() * (num_image_patches - 1)) + sequence_length
+        batch_indices, non_image_indices = torch.where(input_ids != self.config.image_token_index)
+        modality_indicators = torch.zeros((batch_size, max_embed_dim), dtype=torch.long, device=inputs_embeds.device)
```

Comment:
Hi Raushan! For now, my code works with the old version of `merge_input_ids_with_image_features` without any bugs. Also, since I am dealing with Multiway Attention, it needs to keep track of the order and structure of how image features are merged with input ids, more specifically it returns a tensor containing 1s for the image embeds and 0s for the text.

Reply:
Ah, I see. The maintaining order and structure of image/text inputs sounds very similar to what token type ids usually do, iiuc. In that case it will not be hard to integrate it in processing, where it usually belongs. As long as we know what will be the length occupied by an image during processing, we can expand the text with dummy `image tokens` and prepare `token_type_ids` from that by checking if it's a dummy image token or not. 

My biggest concern is that we'll ship a model with a method that is under deprecation, which will slow down the general VLM refactor process. I understand that the PR is not merged but I am now pushing all new VLMs to use the expanded inputs (e.g. Chameleon)